spring-cloud-dataflow-admin/src/main/java/org/springframework/cloud/dataflow/admin/controller/StreamController.java

```diff
@@ -222,34 +222,77 @@ private void deployStream(StreamDefinition stream, Map<String, String> cumulated
 		if (cumulatedDeploymentProperties == null) {
 			cumulatedDeploymentProperties = Collections.emptyMap();
 		}
-		Iterator<ModuleDefinition> iterator = stream.getDeploymentOrderIterator();
-		int nextModuleCount = 0;
-		boolean isDownStreamModulePartitioned = false;
-		for (int i = 0; iterator.hasNext(); i++) {
+
+		ModulePartitionInfo downstreamInfo = ModulePartitionInfo.initial();
+
+		for (StreamDefinition.DeploymentOrderIterator iterator = stream.getDeploymentOrderIterator();
+			iterator.hasNext(); ) {
 			ModuleDefinition currentModule = iterator.next();
-			boolean isFirst = !iterator.hasNext();
-			boolean isLast = (i == 0);
-			ModuleType type = determineModuleType(currentModule, isFirst, isLast);
-			ModuleRegistration registration = this.registry.find(currentModule.getName(), type);
-			if (registration == null) {
-				throw new IllegalArgumentException(String.format(
-						"Module %s of type %s not found in registry", currentModule.getName(), type));
-			}
-			ModuleCoordinates coordinates = registration.getCoordinates();
+
+			ModuleCoordinates coordinates = getModuleCoordinates(iterator, currentModule);
```

Comment:
should not pass the `iterator` itself... perhaps separate the `determineType` method, e.g.:

```
ModuleType type = determineType(currentModule, iterator.hasMoreUpstream(), iterator.hasMoreDownstream());
ModuleCoordinates coordinates = getModuleCoordinates(currentModule, type);
```